src/Illuminate/Database/Eloquent/Concerns/HasCasts.php

```diff
@@ -0,0 +1,349 @@
+
+    /**
+     * Determine if the cast type is a custom date time cast.
+     *
+     * @param  string  $cast
+     * @return bool
+     */
+    protected function isCustomDateTimeCast($cast)
+    {
+        return strncmp($cast, 'date:', 5) === 0 ||
+            strncmp($cast, 'datetime:', 9) === 0;
+    }
+
+    /**
+     * Determine if the cast type is a decimal cast.
+     *
+     * @param  string  $cast
+     * @return bool
+     */
+    protected function isDecimalCast($cast)
+    {
+        return strncmp($cast, 'decimal:', 8) === 0;
+    }
+
+    /**
+     * Determine if the given attribute is a date or date castable.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isDateAttribute($key)
+    {
+        return in_array($key, $this->getDates(), true) ||
+            $this->isDateCastable($key);
+    }
+
+    /**
+     * Is the checked value a custom Cast.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isCustomCastable($key)
+    {
+        if ($cast = $this->getCast($key)) {
+            return is_subclass_of($cast, Castable::class);
+        }
+
+        return false;
+    }
+
+    /**
+     * Determine whether a value is Date / DateTime castable for inbound manipulation.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isDateCastable($key)
+    {
+        return $this->hasCast($key, ['date', 'datetime']);
+    }
+
+    /**
+     * Determine whether a value is JSON castable for inbound manipulation.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isJsonCastable($key)
+    {
+        return $this->hasCast($key, ['array', 'json', 'object', 'collection']);
+    }
+
+    /**
+     * Getting the execution result from a user Cast object.
+     *
+     * @param  string  $key
+     * @param  mixed  $value
+     * @return mixed
+     *
+     * @throws \Illuminate\Contracts\Container\BindingResolutionException
+     */
+    protected function fromCustomCastable($key, $value = null)
+    {
+        return $this
+            ->normalizeHandlerToCallable($key)
+            ->get($value);
+    }
+
+    /**
+     * Converting a value by custom Cast.
+     *
+     * @param  string  $key
+     * @param  mixed  $value
+     * @return mixed
+     *
+     * @throws \Illuminate\Contracts\Container\BindingResolutionException
+     */
+    protected function toCustomCastable($key, $value = null)
+    {
+        return $this
+            ->normalizeHandlerToCallable($key)
+            ->set($value);
+    }
+
+    /**
+     * Getting a custom cast instance.
+     *
+     * @param  string  $key
+     * @return \Illuminate\Contracts\Database\Eloquent\Castable
+     *
+     * @throws \Illuminate\Contracts\Container\BindingResolutionException
+     */
+    protected function normalizeHandlerToCallable($key)
+    {
+        return Container::getInstance()
+            ->make($this->getCast($key));
```

Comment:
This is the code which creates the actual cast object:
1. This means _every time_ a custom cast attribute is processed, at least `make` is called.
2. If you failed to register your castable class as singletons, this will create an instance every time too.

I think both things are not good for performance.

Going through `make` every time => `make` does a lot and can be slow with a lot of custom cast attributes. I consider "model attribute access" a hot path in any Laravel application and custom casts implemented like this will make them not a practical choice (or at least: have a hidden performance cost).

I think something similar to _the concept of_ `\Illuminate\Database\Eloquent\Concerns\HasAttributes::$mutatorCache` should be used.

Or maybe go even further, like Facades: Facades are by default effectively singletons, maybe this should be made a property of the custom casts too. Their design should be stateless from the beginning.

Or even more controversial, simple don't use make but just `new` and cache the instance. Let implementors for complex casts with dependencies deal on their own but make it safe to use for the 99% use case.

Reply:
I agree. I will think about it and make corrections.

Reply:
It is resolved. I thought it would be right to add a variable while storing the initialized classes in it. Upon repeated call, the existence will be checked and either the launched instance will be returned, or the new one will be initialized with saving to the array.